deeppavlov/skills/ecommerce_bot/tfidf_bot.py

```diff
@@ -0,0 +1,275 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import Counter
+from typing import List, Tuple, Dict, Union
+from operator import itemgetter
+from scipy.stats import entropy
+from scipy.sparse import csr_matrix, vstack
+from scipy.sparse.linalg import norm as sparse_norm
+
+import numpy as np
+
+from deeppavlov.core.common.registry import register
+from deeppavlov.core.common.log import get_logger
+from deeppavlov.core.common.file import save_pickle, load_pickle
+from deeppavlov.core.commands.utils import expand_path
+from deeppavlov.core.models.estimator import Component
+
+log = get_logger(__name__)
+
+
+@register("ecommerce_tfidf_bot")
+class EcommerceTfidfBot(Component):
+    """Class to retrieve product items from `load_path` catalogs
+    in sorted order according to the similarity measure
+    Retrieve the specification attributes with corresponding values
+    in sorted order according to entropy.
+
+    Parameters:
+        preprocess: text preprocessing component
+        save_path: path to save a model
+        load_path: path to load a model
+        entropy_fields: the specification attributes of the catalog items
+        min_similarity: similarity threshold for ranking
+        min_entropy: min entropy threshold for specifying
+    """
+
+    def __init__(self,
+                 save_path: str,
+                 load_path: str,
+                 entropy_fields: list,
+                 min_similarity: float = 0.5,
+                 min_entropy: float = 0.5,
+                 **kwargs) -> None:
+
+        self.save_path = expand_path(save_path)
+        self.load_path = expand_path(load_path)
+
+        self.min_similarity = min_similarity
+        self.min_entropy = min_entropy
+        self.entropy_fields = entropy_fields
+        self.ec_data: List = []
+        self.x_train_features = None
+        if kwargs.get('mode') != 'train':
+            self.load()
+
+    def fit(self, data, query) -> None:
+        """Preprocess items `title` and `description` from the `data`
+
+        Parameters:
+            data: list of catalog items
+
+        Returns:
+            None
+        """
+
+        self.x_train_features = vstack(list(query))
+        self.ec_data = data
+
+    def save(self) -> None:
+        """Save classifier parameters"""
+        log.info("Saving to {}".format(self.save_path))
+        path = expand_path(self.save_path)
+        save_pickle((self.ec_data, self.x_train_features), path)
+
+    def load(self) -> None:
+        """Load classifier parameters"""
+        log.info("Loading from {}".format(self.load_path))
+        self.ec_data, self.x_train_features = load_pickle(
+            expand_path(self.load_path))
+
+    def __call__(self, q_vects, histories, states):
```

Comment:
Type annotation

Reply:
If a return type is complex, it doesn't mean you can skip its specification!

Reply:
Since dunder-call method is one of the central in DeepPavlov, pay attention to it at first. Docstrings for private methods are mostly not for rendering on docs pages, so they have lower priority.